crates/net/rpc/src/engine/mod.rs

```diff
@@ -0,0 +1,106 @@
+use crate::result::rpc_err;
+use async_trait::async_trait;
+use jsonrpsee::core::RpcResult as Result;
+use reth_consensus::engine::{EngineApiError, EngineApiResult, EngineMessage};
+use reth_interfaces::consensus::ForkchoiceState;
+use reth_primitives::H64;
+use reth_rpc_api::EngineApiServer;
+use reth_rpc_types::engine::{
+    ExecutionPayload, ForkchoiceUpdated, PayloadAttributes, PayloadStatus, TransitionConfiguration,
+};
+use tokio::sync::{
+    mpsc::UnboundedSender,
+    oneshot::{self, Receiver},
+};
+
+/// The server implementation of Engine API
+pub struct EngineApi {
+    /// Handle to the consensus engine
+    engine_tx: UnboundedSender<EngineMessage>,
+}
+
+impl std::fmt::Debug for EngineApi {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("EngineApi").finish_non_exhaustive()
+    }
+}
+
+impl EngineApi {
+    async fn delegate_request<T>(
+        &self,
+        msg: EngineMessage,
+        rx: Receiver<EngineApiResult<T>>,
+    ) -> Result<T> {
+        let _ = self.engine_tx.send(msg);
+        rx.await.expect("channel is open").map_err(|err| {
+            let code = match err {
+                EngineApiError::PayloadUnknown => -38001,
+                // Any other server error
+                _ => jsonrpsee::types::error::INTERNAL_ERROR_CODE,
+            };
```

Comment:
let's make this a function of `EngineApiError`

I guess then we can also do a From impl For RpcErr?

Reply:
it was like that initially, but after moving to `consensus` crate i didn't want to introduce `jsonrpsee` dep there. perhaps should be a separate crate after all